app/src/main/java/org/stepic/droid/core/presenters/LoginPresenter.kt

```diff
@@ -17,48 +24,103 @@ import org.stepic.droid.util.AppConstants
+            .subscribeOn(backgroundScheduler)
+            .observeOn(mainScheduler)
+            .subscribeBy(
+                onSuccess = { handleSuccess(it, authInfo) },
+                onError = { handleError(it, authInfo)}
+            )
+        // doRequest(api.authWithLoginPassword(login, rawPassword), AuthInfo(type = Type.LOGIN_PASSWORD, credentials = Credentials(login, rawPassword), credential = credential, isAfterRegistration = isAfterRegistration))
     }
 
     fun loginWithCode(rawCode: String) {
+        view?.onLoadingWhileLogin()
         val code = rawCode.trim()
-        doRequest(api.authWithCode(code), AuthInfo(type = Type.SOCIAL, socialType = authSocialType))
+        val authInfo = AuthInfo(type = Type.SOCIAL, socialType = authSocialType)
+        compositeDisposable += authRepository.authWithCode(code)
+            .subscribeOn(backgroundScheduler)
+            .observeOn(mainScheduler)
+            .subscribeBy(
+                onSuccess = { handleSuccess(it, authInfo) },
+                onError = { handleError(it, authInfo)}
+            )
+//        doRequest(api.authWithCode(code), AuthInfo(type = Type.SOCIAL, socialType = authSocialType))
     }
 
     fun loginWithNativeProviderCode(nativeCode: String, type: SocialManager.SocialType, email: String? = null) {
+        view?.onLoadingWhileLogin()
         val code = nativeCode.trim()
-        doRequest(api.authWithNativeCode(code, type, email), AuthInfo(type = Type.SOCIAL, socialType = type))
+        val authInfo = AuthInfo(type = Type.SOCIAL, socialType = type)
+        compositeDisposable += authRepository.authWithNativeCode(code, type, email)
+            .subscribeOn(backgroundScheduler)
+            .observeOn(mainScheduler)
+            .subscribeBy(
+                onSuccess = { handleSuccess(it, authInfo) },
+                onError = { handleError(it, authInfo)}
+            )
+//        doRequest(api.authWithNativeCode(code, type, email), AuthInfo(type = Type.SOCIAL, socialType = type))
+    }
+
+    fun signUp(firstName: String, lastName: String, email: String, password: String) {
+        compositeDisposable += authRepository.createAccount(RegistrationCredentials(firstName, lastName, email, password))
+            .subscribeOn(backgroundScheduler)
+            .observeOn(mainScheduler)
+            .subscribeBy(
+                onComplete = { login(email, password, isAfterRegistration = true)},
+                onError = {
+                    val responseBody = if (it is HttpException) {
+                        it.response()?.errorBody()
+                    } else {
+                        null
+                    }
+                    view?.onRegistrationFailed(responseBody)
+                }
+            )
     }
 
     @MainThread
-    private fun doRequest(callToServer: Call<AuthenticationStepikResponse>, authInfo: AuthInfo) {
+    private fun doRequest(callToServer: Call<OAuthResponse>, authInfo: AuthInfo) {
```

Comment:
Why don't just pass here `source: Single<OAuthResponse>` and change much less code?